vertical-pod-autoscaler/enhancements/3913-customized-recomender-vpa.md

```diff
@@ -0,0 +1,272 @@
+  // default values.
+  // +optional
+  UpdatePolicy *PodUpdatePolicy `json:"updatePolicy,omitempty" protobuf:"bytes,2,opt,name=updatePolicy"`
+
+  // Controls how the autoscaler computes recommended resources.
+  // The resource policy may be used to set constraints on the recommendations
+  // for individual containers. If not specified, the autoscaler computes recommended
+  // resources for all containers in the pod, without additional constraints.
+  // +optional
+  ResourcePolicy *PodResourcePolicy `json:"resourcePolicy,omitempty" protobuf:"bytes,3,opt,name=resourcePolicy"`
+}
+```
+
+When creating a recommender object for recommendations, the recommender main routine should initiate itself 
+with a predefined recommender name, which can be defined as a constant in the `pkg/recomender/main.go` routine,
+
+```golang
+const RecommenderName = "default"
+
+recommender := routines.NewRecommender(config, *checkpointsGCInterval, useCheckpoints, RecommenderName, *vpaObjectNamespace)
+```
+
+where the routines.NewRecommender can pass the `RecommenderName` to the clusterState object.
+
+```golang
+// NewRecommender creates a new recommender instance.
+// Dependencies are created automatically.
+// Deprecated; use RecommenderFactory instead.
+func NewRecommender(config *rest.Config, checkpointsGCInterval time.Duration, useCheckpoints bool, recommender_name string, namespace string) Recommender {
+  clusterState := model.NewClusterState(recommender_name)
+  return RecommenderFactory{
+     ClusterState:           clusterState,
+     ClusterStateFeeder:     input.NewClusterStateFeeder(config, clusterState, *memorySaver, namespace),
+     CheckpointWriter:       checkpoint.NewCheckpointWriter(clusterState, vpa_clientset.NewForConfigOrDie(config).AutoscalingV1()),
+     VpaClient:              vpa_clientset.NewForConfigOrDie(config).AutoscalingV1(),
+     PodResourceRecommender: logic.CreatePodResourceRecommender(),
+     CheckpointsGCInterval:  checkpointsGCInterval,
+     UseCheckpoints:         useCheckpoints,
+  }.Make()
+}
+
+
+// NewClusterState returns a new ClusterState with no pods.
+func NewClusterState(recommender_name string) *ClusterState {
+  return &ClusterState{
+     RecommenderName:   recommender_name,
+     Pods:              make(map[PodID]*PodState),
+     Vpas:              make(map[VpaID]*Vpa),
+     EmptyVPAs:         make(map[VpaID]time.Time),
+     aggregateStateMap: make(aggregateContainerStatesMap),
+     labelSetMap:       make(labelSetMap),
+  }
+}
+```
+
+Therefore, when loading VPA objects to the `clusterStateFeeder`, it can use the field selector to select VPA CRDs that 
+have `recommenderName` equal to the current clusterState’s `RecommenderName`.
+```golang
+// Fetch VPA objects and load them into the cluster state.
+func (feeder *clusterStateFeeder) LoadVPAs() {
+  // List VPA API objects.
+  allVpaCRDs, err := feeder.vpaLister.List(labels.Everything())
```

Comment:
Is it possible to list only the VPA objects that match condition in line 200?

Reply:
I think it should be possible to use selector by field values, but needs to be checked.